src/dbnode/storage/bootstrap/bootstrapper/base.go

```diff
@@ -76,133 +74,261 @@ func (b baseBootstrapper) String() string {
+		b.log.Info("bootstrapping from source completed with error", errorLogFields...)
+	}
+	if err != nil {
+		return bootstrap.NamespaceResults{}, err
 	}
 
-	wg.Wait()
-	if err := xerrors.FirstError(currErr, nextErr); err != nil {
-		return err
+	// Determine the unfulfilled and the unattempted ranges to execute next.
+	next := bootstrap.Namespaces{
+		Namespaces: bootstrap.NewNamespacesMap(bootstrap.NamespacesMapOptions{}),
 	}
+	for _, elem := range namespaces.Namespaces.Iter() {
+		id := elem.Key()
+		namespace := elem.Value()
+
+		currResult, ok := currResults.Results.Get(id)
+		if !ok {
+			return bootstrap.NamespaceResults{},
+				fmt.Errorf("namespace result not returned by bootstrapper: %v", id.String())
+		}
+
+		currNamespace, ok := curr.Namespaces.Get(id)
+		if !ok {
+			return bootstrap.NamespaceResults{},
+				fmt.Errorf("namespace prepared request not found: %v", id.String())
+		}
+
+		// Shallow copy the current namespace for the next namespace prepared request.
+		nextNamespace := currNamespace
+
+		// Calculate bootstrap time ranges.
+		dataRequired := namespace.DataRunOptions.ShardTimeRanges.Copy()
+		dataCurrRequested := currNamespace.DataRunOptions.ShardTimeRanges.Copy()
+		dataCurrFulfilled := dataCurrRequested.Copy()
+		dataCurrFulfilled.Subtract(currResult.DataResult.Unfulfilled())
+
+		dataUnfulfilled := dataRequired.Copy()
+		dataUnfulfilled.Subtract(dataCurrFulfilled)
+
+		// Modify the unfulfilled result.
+		currResult.DataResult.SetUnfulfilled(dataUnfulfilled.Copy())
+
+		// Set the next bootstrapper required ranges.
+		nextNamespace.DataRunOptions.ShardTimeRanges = dataUnfulfilled.Copy()
+
+		var (
+			indexRequired      = result.ShardTimeRanges{}
+			indexCurrRequested = result.ShardTimeRanges{}
+			indexCurrFulfilled = result.ShardTimeRanges{}
+		)
+		if namespace.Metadata.Options().IndexOptions().Enabled() {
+			// Calculate bootstrap time ranges.
+			indexRequired = namespace.IndexRunOptions.ShardTimeRanges.Copy()
+			indexCurrRequested = currNamespace.IndexRunOptions.ShardTimeRanges.Copy()
+			indexCurrFulfilled = indexCurrRequested.Copy()
+			indexCurrFulfilled.Subtract(currResult.IndexResult.Unfulfilled())
+
+			indexUnfulfilled := indexRequired.Copy()
+			indexUnfulfilled.Subtract(indexCurrFulfilled)
+
+			// Modify the unfulfilled result.
+			currResult.IndexResult.SetUnfulfilled(indexUnfulfilled.Copy())
 
-	fulfilledRanges := result.ShardTimeRanges{}
-	fulfilledRanges.AddRanges(currStatus.fulfilled)
-	fulfilledRanges.AddRanges(nextStatus.fulfilled)
-	unfulfilled := totalRanges.Copy()
-	unfulfilled.Subtract(fulfilledRanges)
+			// Set the next bootstrapper required ranges.
+			nextNamespace.IndexRunOptions.ShardTimeRanges = indexUnfulfilled.Copy()
+		} else {
+			// NB(r): Make sure to always set an empty requested range so IsEmpty
+			// does not cause nil ptr deref.
+			nextNamespace.IndexRunOptions.ShardTimeRanges = result.ShardTimeRanges{}
```

Comment:
nit: can simplify by removing the `else` and just setting this to

`nextNamespace.IndexRunOptions.ShardTimeRanges = indexUnfulfilled`

Reply:
Sure thing.